File: SenLib/Sen3/FileFixes/t_name.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	internal class NameData {
		public ushort idx;
		public string name;
		public string str2;
		public string str3;
		public string str4;
		public string str5;
		public string str6;
		public byte[] d2;

		internal NameData(byte[] data) {
			var stream = new DuplicatableByteArrayStream(data);
			idx = stream.ReadUInt16();
			name = stream.ReadUTF8Nullterm();
			str2 = stream.ReadUTF8Nullterm();
			str3 = stream.ReadUTF8Nullterm();
			str4 = stream.ReadUTF8Nullterm();
			str5 = stream.ReadUTF8Nullterm();
			str6 = stream.ReadUTF8Nullterm();
			d2 = stream.ReadBytes(stream.Length - stream.Position);
		}

		internal byte[] ToBinary() {
			MemoryStream ms = new MemoryStream();
			ms.WriteUInt16(idx);
			ms.WriteUTF8Nullterm(name);
			ms.WriteUTF8Nullterm(str2);
			ms.WriteUTF8Nullterm(str3);
			ms.WriteUTF8Nullterm(str4);
			ms.WriteUTF8Nullterm(str5);
			ms.WriteUTF8Nullterm(str6);
			ms.Write(d2);
			return ms.CopyToByteArrayAndDispose();
		}
	}

	class t_name : FileMod {
		public string GetDescription() {
			return "Fix incorrect NPC names.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4ddf1af1f31eeb6cul, 0x9311804d88b3f779ul, 0x519df457u));
			if (file_en == null) {
				return null;
			}
			var tbl_en = new Tbl(file_en, EndianUtils.Endianness.LittleEndian);

			// Tally -> Tallys (Zero/Azure/CS4 consistency)
			{
				var m = new NameData(tbl_en.Entries[595].Data);
				m.name += 's';
				tbl_en.Entries[595].Data = m.ToBinary();
			}

			// Second Lieutenant Midrun -> 2nd Lieutenant Midrun (CS4 consistency)
			{
				var m = new NameData(tbl_en.Entries[622].Data);
				m.name = m.name.ReplaceSubstring(0, 4, "2", 0, 1);
				tbl_en.Entries[622].Data = m.ToBinary();
			}

			// Second Lieutenant Sark -> 2nd Lieutenant Sark (CS4 consistency)
			{
				var m = new NameData(tbl_en.Entries[623].Data);
				m.name = m.name.ReplaceSubstring(0, 4, "2", 0, 1);
				tbl_en.Entries[623].Data = m.ToBinary();
			}

			Stream result_en = new MemoryStream();
			tbl_en.WriteToStream(result_en, EndianUtils.Endianness.LittleEndian);
			DuplicatableByteArrayStream result_en_b = result_en.CopyToByteArrayStream();
			return new FileModResult[] {
				new FileModResult("data/text/dat_en/t_name.tbl", result_en_b.Duplicate()),
			};
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4ddf1af1f31eeb6cul, 0x9311804d88b3f779ul, 0x519df457u));
			if (file_en == null) {
				return null;
			}
			return new FileModResult[] {
				new FileModResult("data/text/dat_en/t_name.tbl", file_en.Duplicate()),
			};
		}
	}
}

File: native/sen3/FileFixes/t_name.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen3/file_fixes.h"
#include "sen3/tbl.h"
#include "sha1.h"
#include "util/text.h"

namespace SenLib::Sen3::FileFixes::t_name {
std::string_view GetDescription() {
    return "Fix incorrect NPC names.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    using namespace HyoutaUtils::TextUtils;

    try {
        auto file = getCheckedFile(
            "data/text/dat_en/t_name.tbl",
            150070,
            SenPatcher::SHA1FromHexString("4ddf1af1f31eeb6c9311804d88b3f779519df457"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;
        Tbl tbl_en(bin.data(), bin.size(), HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // Tally -> Tallys (Zero/Azure/CS4 consistency)
        {
            auto& e = tbl_en.Entries[595];
            NameData m(e.Data.data(), e.Data.size());
            m.name += 's';
            e.Data = m.ToBinary();
        }

        // Second Lieutenant Midrun -> 2nd Lieutenant Midrun (CS4 consistency)
        {
            auto& e = tbl_en.Entries[622];
            NameData m(e.Data.data(), e.Data.size());
            m.name = ReplaceSubstring(m.name, 0, 4, "2", 0, 1);
            e.Data = m.ToBinary();
        }

        // Second Lieutenant Sark -> 2nd Lieutenant Sark (CS4 consistency)
        {
            auto& e = tbl_en.Entries[623];
            NameData m(e.Data.data(), e.Data.size());
            m.name = ReplaceSubstring(m.name, 0, 4, "2", 0, 1);
            e.Data = m.ToBinary();
        }


        std::vector<char> result_en_vec;
        MemoryStream result_en(result_en_vec);
        tbl_en.WriteToStream(result_en, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(result_en_vec), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::t_name
